### PR TITLE
Openlaw object WIP

### PR DESCRIPTION
**Pull request by Jdville03**
- [x] Structure and organization of methods for Openlaw.scala
- [x] Fill in TODO descriptions
- [x] Examples

### CODE REVIEW

File: docs/openlaw-object/README.md [hunk 1]
```diff
@@ -0,0 +1,1857 @@
...
+
+```
+
+```
+
+### execute
+
+TODO.
+
+```
```
Comment (mroth) [inline docs/openlaw-object/README.md]: VuePress can highlight most popular language syntax, so if you change the start of the block to hint language via:

    ```js

Then it should highlight! Yay!

Comment (Jdville03) [reply docs/openlaw-object/README.md]: oh nice. thanks!

File: docs/openlaw-object/README.md [hunk 2]
```diff
@@ -0,0 +1,2141 @@
...
+  Years Vesting: "4",
+  Unit of Vesting: "250",
+  ...
+};
+Openlaw.executeForReview(compiledTemplate.compiledTemplate, signatures, {}, params);
+```
+
+**Response**
+
+Returns an object containing a [`TemplateExecutionResult` object](#templateexecutionresult) (and key/value pairs indicating that no error has occurred and that no templates are missing) if execution is successful. The [`TemplateExecutionResult` object](#templateexecutionresult) contains a nested object with the ID and name of each signatory as a key/value pair.
```
Comment (outkaj) [inline docs/openlaw-object/README.md]: I am not sure about the ordering here. It might be clearer to explain what the `TemplateExecutionResult` object is before talking about the key/value pairs which indicate errors have occurred.

File: docs/openlaw-object/README.md [hunk 3]
```diff
@@ -0,0 +1,2141 @@
...
+{
+  isError: false,
+  errorMessage: "",
+  compiledTemplate: CompiledTemplate
+}
+```
+
+### execute
+
+Execute (load and run) a compiled template.
```
Comment (outkaj) [inline docs/openlaw-object/README.md]: Should we explain a bit more what it means to load and run a template or are we assuming that information will be familiar to someone who is reading?

File: docs/openlaw-object/README.md [hunk 4]
```diff
@@ -0,0 +1,2141 @@
...
+  {
+    name: "Number of Shares",
+    value: "1000"
+  }
+]
+```
+
+### validateContract
+
+Validate an executed template returned from the [`execute` method](#execute) or [`executeForReview` method](#executeforreview).
```
Comment (outkaj) [inline docs/openlaw-object/README.md]: Should we explain more what it means to validate the template / why validation is needed? Perhaps also something further regarding the transition from template to contract?

File: docs/openlaw-object/README.md [hunk 5]
```diff
@@ -0,0 +1,2141 @@
...
+
+Example
+
+```js
+"Advisor Agreement"
+```
+
+### getAgreements
+
+Get all agreements linked to a template for rendering. A [deal](/markup-language/#deals) template will have one or more agreements. A non-deal template will have only one agreement.
```
Comment (outkaj) [inline docs/openlaw-object/README.md]: Is there a situation where a deal would have just one agreement or is it always more than one? If not, then -> A deal will have more than one agreement

File: docs/openlaw-object/README.md [hunk 6]
```diff
@@ -0,0 +1,2141 @@
...
+
+Example
+
+```html
+<div class='openlaw-paragraph paragraph-1'><p class='no-section'>This Advisor Agreement is entered into between [[Company Name]] ("Corporation") and John Smith ("Advisor") as of September 19, 2018 ("Effective Date"). The parties agree as follows:</p></div><ul class='list-lvl-1'><li><div class='openlaw-paragraph paragraph-2'><p>1.  <strong>Services</strong>. Advisor agrees to consult with and advise Company from time to time, at Company's request (the "Services"). </p></div></li>
+```
+
+### parseMarkdown
+
+Parse selected paragraph text to edit (in draft mode) to HTML.
```
Comment (outkaj) [inline docs/openlaw-object/README.md]: Should we explain a bit more about the paragraphs? Is there anything relevant we can link to from the markup language docs?

File: docs/openlaw-object/README.md [hunk 7]
```diff
@@ -0,0 +1,2141 @@
...
+
+Example
+
+```html
+<p class='no-section'>This Advisor Agreement is entered into between [[Company Name]] ("Corporation") and John Smith ("Advisor") as of September 19, 2018 ("Effective Date"). The parties agree as follows:</p>
+```
+
+### getTypes
+
+Get all [markup language](/markup-language/) variable types for use in markdown editor.
```
Comment (outkaj) [inline docs/openlaw-object/README.md]: Is it confusing how we are using the terms markup vs markdown? Is there any inconsistency there?

File: docs/openlaw-object/README.md [hunk 8]
```diff
@@ -0,0 +1,2141 @@
...
+```js
+[
+  "Collection",
+  "Address",
+  "Choice",
+  "Date",
+  "DateTime",
+  "EthAddress",
+  "EthereumCall",
+  "StripeCall",
```
Comment (outkaj) [inline docs/openlaw-object/README.md]: Is the issue with removing the `StripeCall` here that technically it will be returned in the response (so it's not accurate to remove it)?

File: docs/openlaw-object/README.md [hunk 9]
```diff
@@ -0,0 +1,2141 @@
...
+  "Identity",
+  "Number",
+  "Period",
+  "SmartContractMetadata",
+  "Structure",
+  "Template",
+  "Text",
+  "Validation",
+  "YesNo"
```
Comment (outkaj) [inline docs/openlaw-object/README.md]: Is there a way we could link to these types' descriptions in the markup language docs? Or is that unnecessary?

File: docs/openlaw-object/README.md [hunk 10]
```diff
@@ -0,0 +1,2141 @@
...
+
+**Response**
+
+Returns `true` if template is a [deal](/markup-language/#deals) template.
+
+## Variable
+
+### showInForm
+
+Check if variable is shown in the input form in draft mode.
```
Comment (outkaj) [inline docs/openlaw-object/README.md]: Let's explain a bit more what this means.

File: docs/openlaw-object/README.md [hunk 11]
```diff
@@ -0,0 +1,2141 @@
...
+
+Returns name of variable as a string.
+
+Example
+
+```js
+"Company Name"
+"Effective Date"
+"No Services"
+"Number of Shares"
```
Comment (outkaj) [inline docs/openlaw-object/README.md]: I think it is clearer to return just one string, otherwise it looks like an array without brackets (unless I am missing something)

File: docs/openlaw-object/README.md [hunk 12]
```diff
@@ -0,0 +1,2141 @@
...
+const executionResult = Openlaw.execute(compiledTemplate.compiledTemplate, {}, params);
+const allVariables = Openlaw.getVariables(executionResult.executionResult, {});
+allVariables.forEach(variable => {
+  Openlaw.getDescription(variable));
+});
+```
+
+**Response**
+
+Returns description of variable as a string. The default description is the variable name. You can also modify the description that appears in the form as described in our [markup language documentation](/markup-language/#variables).
```
Comment (outkaj) [inline docs/openlaw-object/README.md]: The default description is the variable name. You can also modify the description that appears in the form as described in our [markup language documentation](/markup-language/#variables). -> I think it would be clearer if this was included after "Get description of a variable" above.

File: docs/openlaw-object/README.md [hunk 13]
```diff
@@ -0,0 +1,2141 @@
...
+```
+
+**Response**
+
+Returns an array of option values for the [Choice type](/markup-language/#choice) as strings.
+
+Example
+
+```js
+[
```
Comment (outkaj) [inline docs/openlaw-object/README.md]: :)

File: docs/openlaw-object/README.md [hunk 14]
```diff
@@ -0,0 +1,2141 @@
...
+
+Example
+
+```js
+// see examples above for #execute and #executeForReview for parameters
+const executionResult = Openlaw.execute(compiledTemplate.compiledTemplate, {}, params);
+const allVariables = Openlaw.getVariables(executionResult.executionResult, {});
+const variable = allVariables[0]; // first variable in template is a Structure type
+const field = Openlaw.getStructureFieldDefinitions(variable, executionResult.executionResult)[0]; // variable being checked is first variable field in Structure type variable
+const structureValue = '{"First name":"John","Last
```
Comment (outkaj) [inline docs/openlaw-object/README.md]: Let's include this Shared Types section at the top instead since it provides useful context for what follows.

Comment (outkaj) [inline docs/openlaw-object/README.md]: I think it could be useful to have headings for (for example) methods on collections, methods on structure types, etc...

**Comment by outkaj** [comment]
@Jdville03 this looks great! Very thorough. I left only a few small comments and questions. (Also, nice easter egg re: country list :))

File: docs/openlaw-object/README.md (hunk 5 above)
Comment (Jdville03) [reply docs/openlaw-object/README.md]: It's possible for a deal template to be linked to just one agreement. For example, the Employee Onboarding deal could end just having the Employee Offer Letter because the other agreements are conditional. Also, technically you can set up a deal to just include one agreement (even though the usefulness of a deal is bundling multiple agreements).

File: docs/openlaw-object/README.md [hunk 15]
```diff
@@ -0,0 +1,2141 @@
+# Openlaw Object
+
+The `Openlaw` object defined in Openlaw.scala compiles to JavaScript and is an interface in the OpenLaw protocol to interact directly with an agreement and its contents, including its various variable types. The object methods are categorized below.
```
[the remark this answers is not in the document]
Comment (Jdville03) [reply docs/openlaw-object/README.md]: @outkaj @craigwblake-openlaw @adridadou - is there a short blurb you can provide to help explain ScalaJS a little more here?

File: docs/openlaw-object/README.md (hunk 3 above)
Comment (Jdville03) [reply docs/openlaw-object/README.md]: @adridadou what are your thoughts on this one? Is there something more we can add here? I included "load and run" to describe "execute" to distinguish from the legal meaning of execute (all the parties signing the contract).

File: docs/openlaw-object/README.md (hunk 4 above)
Comment (Jdville03) [reply docs/openlaw-object/README.md]: @adridadou I have the same questions too. Do you have any suggestions on what we could add here?

File: docs/openlaw-object/README.md (hunk 14 above)
Comment (Jdville03) [reply docs/openlaw-object/README.md]: @outkaj the category headings already include Template, Variable, Address, Choice, Collection, Identity, Structure, and Shared Types. Are you thinking the headings should be different?

File: docs/openlaw-object/README.md (hunk 2 above)
Comment (Jdville03) [reply docs/openlaw-object/README.md]: Changed the order around to make this clearer.

File: docs/openlaw-object/README.md (hunk 6 above)
Comment (Jdville03) [reply docs/openlaw-object/README.md]: I added more of an explanation of editing paragraphs. There is no explanation of it in the markup language docs but I included a link to https://media.consensys.net/frictionless-contracting-628884915b8c which explains our draft tools and includes a demo video showing how to edit paragraphs directly.

File: docs/openlaw-object/README.md (hunk 9 above)
Comment (Jdville03) [reply docs/openlaw-object/README.md]: Probably isn't necessary to link for each if we just provide a link to the markup language docs in explaining what the response is right before this list. I changed to this:

Returns array of input and specialized variable types as strings. Learn more about these different types in our [Markup Language documentation](/markup-language/).

File: docs/openlaw-object/README.md (hunk 10 above)
Comment (Jdville03) [reply docs/openlaw-object/README.md]: I updated the language here with a better explanation. @outkaj and @adridadou can you check if that's sufficient?

File: docs/openlaw-object/README.md (hunk 11 above)
Comment (Jdville03) [reply docs/openlaw-object/README.md]: Removed the code block here and just referred to one example of `"Company Name"`.

File: docs/openlaw-object/README.md (hunk 12 above)
Comment (Jdville03) [reply docs/openlaw-object/README.md]: I agree. I moved the explanation up to right after "Get description of a variable." under the heading.

File: docs/openlaw-object/README.md (hunk 14 above)
Comment (Jdville03) [reply docs/openlaw-object/README.md]: Sure - moved to the top of this page of the docs.

File: docs/README.md [hunk 16]
```diff
@@ -1,6 +1,6 @@
-# Introduction
+# Overview
 
-Welcome to OpenLaw's documentation. Here you'll learn more about OpenLaw, our markup language, and how our first application works. If you have any questions email us at [hello@openlaw.io](mailto:hello@openlaw.io).
+Welcome to OpenLaw's documentation. Here you'll learn more about OpenLaw, including our markup language, our protocol which includes JavaScript APIs for convenient integration, and how our application works in general. If you have any questions, email us at [hello@openlaw.io](mailto:hello@openlaw.io).
```
[the remark this answers is not in the document]
Comment (outkaj) [reply docs/README.md]: great!

File: docs/.vuepress/config.js [hunk 17]
```diff
@@ -7,14 +7,16 @@ module.exports = {
...
     sidebar: [
       "/",
-      "/markup-language/"
+      "/markup-language/",
+      "/api-client/",
+      "/openlaw-object/"
       // add new top level sections here...
     ],
     algolia: {
       apiKey: "68c3c0d36676a3acce1cd3c7087bc7c9",
```
[the remark this answers is not in the document]
Comment (Jdville03) [reply docs/.vuepress/config.js]: @mroth just wanted to bump this as we're getting close to making the protocol docs go live.

[the remark this answers is not in the document]
Comment (mroth) [reply docs/.vuepress/config.js]: Nope. It’s basically just an ID, it’s fine / intended to be publicly readable since it’s on the frontend to query DocSearch.
(Thanks for the ping. Somehow I missed this until now.)

File: docs/.vuepress/public/ol-logo-color.svg [hunk 18]
```diff
@@ -0,0 +1,15 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<svg width="50px" height="50px" viewBox="0 0 50 50" version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink">
+    <!-- Generator: Sketch 51.2 (57519) - http://www.bohemiancoding.com/sketch -->
```
Comment (mroth) [inline docs/.vuepress/public/ol-logo-color.svg]: Running this file through svgo (or the web version at svgomg) is likely to make some minimal size improvements, removing extraneous comments like this and some other no-op markup. [FWIW generally I’d prefer to keep UI changes like this in a different PR so they can be discussed separately, if they don’t have bearing on the content changes.]

File: .vscode/settings.json [hunk 19]
```diff
@@ -8,4 +8,12 @@
     "spellright.ignoreRegExpsByClass": {
         "markdown": ["/:\\w+:/g"]
     },
+    "spellright.language": [
+      "en"
+    ],
+    "spellright.documentTypes": [
+      "latex",
```
Comment (mroth) [inline .vscode/settings.json]: We have LaTeX documents?

File: docs/.vuepress/public/ol-logo-color.svg (hunk 18 above)
Comment (Jdville03) [reply docs/.vuepress/public/ol-logo-color.svg]: Thanks for the tip. Replaced this file after running through svgomg.

File: .vscode/settings.json (hunk 19 above)
Comment (Jdville03) [reply .vscode/settings.json]: latex (with plaintext and markdown) automatically got inserted into the workspace settings when I selected English as the spellcheck language.

File: docs/.vuepress/config.js (hunk 17 above)
[the remark this answers is not in the document]
Comment (Jdville03) [reply docs/.vuepress/config.js]: sounds good. thanks for clarifying

[the remark this answers is not in the document]
Comment (outkaj) [reply docs/.vuepress/config.js]: thanks! glad to hear it

File: docs/openlaw-object/README.md (hunk 14 above)
Comment (outkaj) [reply docs/openlaw-object/README.md]: @Jdville03 missed that, sorry, I was looking just from text and not from deploy preview.

File: docs/openlaw-object/README.md (hunk 5 above)
Comment (outkaj) [reply docs/openlaw-object/README.md]: Ah, I see. Good points!

File: docs/openlaw-object/README.md (hunk 8 above)
Comment (outkaj) [reply docs/openlaw-object/README.md]: If it's technically returned but not mentioned elsewhere then i think it is probably ok to just leave it like this.

File: docs/openlaw-object/README.md (hunk 9 above)
Comment (outkaj) [reply docs/openlaw-object/README.md]: perfect!

File: docs/openlaw-object/README.md (hunk 15 above)
[the remark this answers is not in the document]
Comment (Jdville03) [reply docs/openlaw-object/README.md]: updated with a short explanation and link to Scala.js

[the remark this answers is not in the document]
Comment (outkaj) [reply docs/openlaw-object/README.md]: great! sorry, i don't know why your comment from 3 days ago literally only showed up for me now. marking resolved
